cudax/include/cuda/experimental/__memory_resource/cuda_memory_pool.h

```diff
@@ -0,0 +1,270 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__MEMORY_RESOURCE_CUDA_MEMORY_POOL_H
+#define _CUDAX__MEMORY_RESOURCE_CUDA_MEMORY_POOL_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+// cudaMallocAsync was introduced in CTK 11.2
+#if !defined(_CCCL_COMPILER_MSVC_2017) && !defined(_CCCL_CUDACC_BELOW_11_2)
+
+#  if !defined(_CCCL_CUDA_COMPILER_NVCC) && !defined(_CCCL_CUDA_COMPILER_NVHPC)
+#    include <cuda_runtime.h>
+#    include <cuda_runtime_api.h>
+#  endif // !_CCCL_CUDA_COMPILER_NVCC && !_CCCL_CUDA_COMPILER_NVHPC
+
+#  include <cuda/__memory_resource/get_property.h>
+#  include <cuda/__memory_resource/properties.h>
+#  include <cuda/__memory_resource/resource_ref.h>
+#  include <cuda/std/__cuda/api_wrapper.h>
+#  include <cuda/std/__new_>
+#  include <cuda/stream_ref>
+
+#  include <cuda/experimental/__stream/stream.cuh>
+
+#  if _CCCL_STD_VER >= 2014
+
+namespace cuda::experimental::mr
+{
+
+//! @brief  Checks whether the current device supports cudaMallocAsync
+//! @param __device_id The device id of the current device
+//! @throws cuda_error if cudaDeviceGetAttribute failed
+//! @returns true if cudaDevAttrMemoryPoolsSupported is not zero
+_CCCL_NODISCARD inline bool __device_supports_pools(const int __device_id)
+{
+  int __pool_is_supported = 0;
+  _CCCL_TRY_CUDA_API(
+    ::cudaDeviceGetAttribute,
+    "Failed to call cudaDeviceGetAttribute",
+    &__pool_is_supported,
+    ::cudaDevAttrMemoryPoolsSupported,
+    __device_id);
+  return __pool_is_supported != 0;
+}
+
+//! @brief Internal redefinition of ``cudaMemAllocationHandleType``
+//! @note We need to define our own enum here because the earliest CUDA runtime version that supports asynchronous
+//! memory pools (CUDA 11.2) did not support these flags. See the <a
+//! href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html">cudaMemAllocationHandleType</a> docs
```

Comment:
Do we want this to be doxygen documented? 

```suggestion
//! @brief Internal redefinition of ``cudaMemAllocationHandleType``.
//! @note We need to define our own enum here because the earliest CUDA runtime version that supports asynchronous
//! memory pools (CUDA 11.2) did not support these flags. See the <a
//! href="https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__TYPES.html">cudaMemAllocationHandleType</a> docs.
```